source/extensions/clusters/aggregate/cluster.cc

```diff
@@ -0,0 +1,152 @@
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    int priority_in_current_cluster = 0;
+    for (const auto& host_set : tlc->prioritySet().hostSetsPerPriority()) {
+      if (!host_set->hosts().empty()) {
+        priority_set.updateHosts(
+            next_priority_after_linearizing++, Upstream::HostSetImpl::updateHostsParams(*host_set),
+            host_set->localityWeights(), host_set->hosts(), {}, host_set->overprovisioningFactor());
+        priority_to_cluster.emplace_back(std::make_pair(priority_in_current_cluster, tlc));
+      }
+      priority_in_current_cluster++;
+    }
+  }
+
+  return std::make_pair(std::move(priority_set), std::move(priority_to_cluster));
+}
+
+void Cluster::startPreInit() {
+  for (const auto& cluster : clusters_) {
+    auto tlc = cluster_manager_.get(cluster);
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    // Add callback for clusters initialized before aggregate cluster.
+    tlc->prioritySet().addMemberUpdateCb(
+        [this, cluster](const Upstream::HostVector&, const Upstream::HostVector&) {
+          ENVOY_LOG(info, "member update for cluster '{}' in aggregate cluster '{}'", cluster,
+                    this->info()->name());
+          refresh();
+        });
+  }
+  refresh();
+  handle_ = cluster_manager_.addThreadLocalClusterUpdateCallbacks(*this);
+
+  onPreInitComplete();
+}
+
+void Cluster::refresh(const std::function<bool(const std::string&)>& skip_predicate) {
+  // Post the priority set to worker threads.
+  tls_->runOnAllThreads([this, skip_predicate, cluster_name = this->info()->name()]() {
+    PriorityContext priority_set = linearizePrioritySet(skip_predicate);
+    Upstream::ThreadLocalCluster* cluster = cluster_manager_.get(cluster_name);
+    if (cluster == nullptr) {
+      return;
+    }
+    dynamic_cast<AggregateClusterLoadBalancer&>(cluster->loadBalancer()).refresh(priority_set);
+  });
+}
+
+void Cluster::onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) {
+  if (std::find(clusters_.begin(), clusters_.end(), cluster.info()->name()) != clusters_.end()) {
+    ENVOY_LOG(info, "adding or updating cluster '{}' for aggregate cluster '{}'",
+              cluster.info()->name(), info()->name());
+    refresh();
+    cluster.prioritySet().addMemberUpdateCb(
+        [this](const Upstream::HostVector&, const Upstream::HostVector&) { refresh(); });
+  }
+}
+
+void Cluster::onClusterRemoval(const std::string& cluster_name) {
+  //  The onClusterRemoval callback is called before the thread local cluster was removed. There
+  //  will be a dangling pointer to the thread local cluster if delete cluster is not skipped when
+  //  we refresh the load balancer.
+  if (std::find(clusters_.begin(), clusters_.end(), cluster_name) != clusters_.end()) {
+    ENVOY_LOG(info, "removing cluster '{}' from aggreagte cluster '{}'", cluster_name,
+              info()->name());
+    refresh([cluster_name](const std::string& c) { return cluster_name == c; });
```

Comment:
nit: pass cluster_name by ref since we know it's going to be evaluated inline and not passed to another thread

Reply:
It was by ref before but asan and tsan failed...

Reply:
oh i see, it's because we do pass it to other threads through `runOnAllThreads`. lets keep it as is then